### PR TITLE
DWC I3C master

### PR DESCRIPTION
**Pull request by shashankprashar**
This introduces a basic API for DesignWare I3C master controller. 
This is developed using emulation platform.

Here are a few things that I have tested :

Broadcast RSTDAA works.

Broadcast ENTDAA works.
GETBCR/GETDCR works.
Able to get chip ID from the I3C slave devices.
Dynamic Address assignment tested.
I3C Private R/W working.
I3C CCC Commands works.
I3C SDR and HDR-DDR mode works.

Following things are to be done:
I3C Hot-Join
I3C HDR other modes
I2C Legacy Device Support
Broadcast SETDASA
I3C slave mode support

### CODE REVIEW

File: include/zephyr/drivers/i3c.h [hunk 1]
```diff
@@ -1539,6 +1541,56 @@ static inline int i3c_reg_update_byte(struct i3c_device_desc *target,
...
+ */
+static inline int i3c_ddr_write(struct i3c_device_desc *target,
+			    const uint8_t *buf, uint32_t num_bytes)
+{
+	struct i3c_msg msg;
+
+	msg.buf = (uint8_t *)buf;
+	msg.len = num_bytes;
+	msg.flags = I3C_MSG_WRITE | I3C_MSG_STOP;
+	msg.hdr_mode = 1;
```
Comment (dcpleung) [inline include/zephyr/drivers/i3c.h:1564]: Please set the flags `I3C_MSG_HDR` and `I3C_MSG_HDR_DDR` for mode. Ditto for other similar modifications.

**Comment by nashif** [comment]
would be nice to have this also associated with a platform so we can at least build the code.

File: drivers/i3c/i3c_dw.c [hunk 2]
```diff
@@ -0,0 +1,1743 @@
...
+/**
+ * @brief Get address slot status of the bus
+ *
+ * @param addrslots: Pointer to the address slots info
+ * @param addr: Address of the node
+ *
+ * @retval  Address status value if successful
+ * @retval -EINVAL if address slots not initialized
+ */
+int32_t i3c_bus_get_addr_slot_status(uint32_t *addrslots, uint16_t addr)
```
Comment (XenuIsWatching) [inline drivers/i3c/i3c_dw.c:707]: Any reason not to use the existing functions in `i3c_common.h`?

....although this whole code skimming through looks eerily familiar to me. I've seen this before years ago when it was written for zephyr v2.3 (I think?). It probably needs to now be considered with what landed in v3.2 a bit more with what's in `i3c_common.h` and `i3c_ccc.h` so there is no duplicate code. Was this not known about before or was there a reason to not use these?

File: drivers/i3c/i3c_dw.c [hunk 3]
```diff
@@ -0,0 +1,1743 @@
...
+	.do_ccc = dw_i3c_send_ccc_cmd,
+	.ibi_enable = dw_i3c_enable_ibi,
+	.ibi_disable = dw_i3c_disable_ibi,
+};
+
+#define I3C_DW_IRQ_HANDLER(n)				\
+	static void i3c_dw_irq_config_##n(void)		\
+{							\
+	IRQ_CONNECT(DT_INST_IRQN(n),			\
+			1,				\
```
Comment (XenuIsWatching) [inline drivers/i3c/i3c_dw.c:1704]: just fixed priority? this should be defined in the dts

**Review by XenuIsWatching — CHANGES_REQUESTED** [review]
It's good to see samsung open-sourceing this driver....but think there needs to be more work on integrating it with what landed upstream.

File: include/zephyr/drivers/i3c.h (hunk 1 above)
Comment (shashankprashar) [reply include/zephyr/drivers/i3c.h:1564]: okay will update accordingly.

**Comment by shashankprashar** [comment]
> 

This has been tested in internal emulation environment, as of now we do not have plan to upstream this platform.

**Comment by shashankprashar** [comment]
> in

**Comment by shashankprashar** [comment]
> would be nice to have this also associated with a platform so we can at least build the code.

This has been tested in internal emulation environment, as of now we do not have plan to upstream this platform.

File: drivers/i3c/i3c_dw.c (hunk 3 above)
Comment (shashankprashar) [reply drivers/i3c/i3c_dw.c:1704]: Okay.

File: drivers/i3c/i3c_dw.c (hunk 2 above)
Comment (shashankprashar) [reply drivers/i3c/i3c_dw.c:707]: No specific reason, I am working on utilizing existing APIs and ensuring it does not break any functionality. Will post updated patch soon.

**Comment by shashankprashar** [comment]
@dcpleung @XenuIsWatching I have addressed the review comments and updated the patches please review.

**Review by dcpleung — CHANGES_REQUESTED** [review]
Please update terminology to I3C v1.1.1 so that we can have a more unified presentation (e.g. master -> controller). Note that the documentation is written according to v1.1.1. Using old terminology is going to confuse developers down the road.

**Review by XenuIsWatching — CHANGES_REQUESTED** [review]
another review round for me... but I need to a more deeper dive with how the daa here would attached with an existing driver (such as an imu sensor) when I can get more time

File: drivers/i3c/i3c_dw.c [hunk 4]
```diff
@@ -0,0 +1,1631 @@
...
+
+#include "i3c_dw.h"
+
+LOG_MODULE_REGISTER(i3c_dw);
+
+#define DEV_CTRL(config)	((config)->regs + DEVICE_CTRL)
+#define DEV_CONFIG(dev)		((const struct dw_i3c_config *)(dev)->config)
+#define DEV_DATA(dev)		((struct dw_i3c_data *)(dev)->data)
+#define DIV_ROUND_UP(a, b)	(((a) + (b) - 1) / b)
+#define NUM_ADDR_SLOT		(((I3C_MAX_ADDR + 1) * 2) / BITS_PER_LONG)
```
[the remark this answers is not in the document]
Comment (shashankprashar) [reply drivers/i3c/i3c_dw.c:24]: Missed, ok will remove.

File: drivers/i3c/i3c_dw.h [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+#define I3C_HDR_MODE				(0x1)
+#define I2C_SLAVE				(2)
+#define I3C_SLAVE				(3)
+#define I3C_GETMXDS_FORMAT_1_LEN		(2)
+#define I3C_GETMXDS_FORMAT_2_LEN		(5)
+
+/*
+ * I3C CCC (Common Command Codes) related definitions.
+ */
+#define I3C_CCC_DIRECT				BIT(7)
```
[the remark this answers is not in the document]
Comment (shashankprashar) [reply drivers/i3c/i3c_dw.h:173]: Ok will use from there.

File: drivers/i3c/i3c_dw.h [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+ * CCC event bits.
+ */
+#define I3C_CCC_EVENT_SIR			BIT(0)
+#define I3C_CCC_EVENT_MR			BIT(1)
+#define I3C_CCC_EVENT_HJ			BIT(3)
+
+/*
+ * BCR (Bus Characteristic Register) bits.
+ */
+#define I3C_BCR_I3C_SLAVE			(0x0U << 6)
```
[the remark this answers is not in the document]
Comment (shashankprashar) [reply drivers/i3c/i3c_dw.h:187]: okay

File: drivers/i3c/i3c_dw.c [hunk 7]
```diff
@@ -0,0 +1,1631 @@
...
+ * @brief Get the Provisional ID of slave
+ *
+ * @param dev: Pointer to the device structure
+ * @param info: Pointer to structure containing slave device info
+ *
+ * @retval  0 if successful
+ * @retval -EINVAL if a structure is found NULL
+ * @retval  Otherwise a value linked to transfer errors as per I3C specs
+ */
+static int32_t i3c_master_getpid(const struct device *dev,
```
[the remark this answers is not in the document]
Comment (shashankprashar) [reply drivers/i3c/i3c_dw.c:902]: Okay will use from i3c_ccc.h

File: drivers/i3c/i3c_dw.c [hunk 8]
```diff
@@ -0,0 +1,1631 @@
...
+ * @brief Send CCC command to get maximum data speed
+ *
+ * @param dev: Pointer to the device structure
+ * @param info: Pointer to structure containing slave device info
+ *
+ * @retval  0 if successful
+ * @retval -EINVAL if a structure is found NULL
+ * @retval  Otherwise a value linked to transfer errors as per I3C specs
+ */
+static int32_t i3c_master_getmxds(const struct device *dev,
```
[the remark this answers is not in the document]
Comment (shashankprashar) [reply drivers/i3c/i3c_dw.c:861]: Okay will check and try integrating.

File: drivers/i3c/i3c_dw.c [hunk 9]
```diff
@@ -0,0 +1,1631 @@
...
+ * @brief Get the Device Characteristic Register
+ *
+ * @param dev: Pointer to the device structure
+ * @param info: Pointer to structure containing slave device info
+ *
+ * @retval  0 if successful
+ * @retval -EINVAL if a structure is found NULL
+ * @retval  Otherwise a value linked to transfer errors as per I3C specs
+ */
+static int32_t i3c_master_getdcr(const struct device *dev,
```
[the remark this answers is not in the document]
Comment (shashankprashar) [reply drivers/i3c/i3c_dw.c:962]: okay

**Comment by shashankprashar** [comment]
"you might be able to use i3c_device_basic_info_get for all this. This MAX_DATA_SPEED_LIM could be something to be considered added to that function"
"I'd also expect a GETHDRCAP to be added as well since DDR support is coming with this and for that bit to be checked when attempting a DDR operation"

- Let me check and test this, will get back with updated patch set for same.

**Comment by carlescufi** [comment]
Hey @shashankprashar sorry this PR didn't get in, but now that we have an i3c API in the tree, maybe you could submit your DW driver using the new API?

**Pull request closed without merging**